clients/drcachesim/tests/burst_futex.cpp

```diff
@@ -0,0 +1,196 @@
+            std::cerr << "raw2trace failed: " << error << "\n";
+            assert(false);
+        }
+    }
+    dr_standalone_exit();
+    return outdir;
+}
+
+static std::string
+gather_trace(const std::string &tracer_ops, const std::string &out_subdir)
+{
+    std::string dr_ops("-stderr_mask 0xc -client_lib ';;-offline " + tracer_ops + "'");
+    if (!my_setenv("DYNAMORIO_OPTIONS", dr_ops.c_str()))
+        std::cerr << "failed to set env var!\n";
+    dr_app_setup();
+    assert(!dr_app_running_under_dynamorio());
+    dr_app_start();
+    assert(dr_app_running_under_dynamorio());
+    do_some_work();
+    dr_app_stop_and_cleanup();
+    assert(!dr_app_running_under_dynamorio());
+
+    return post_process(out_subdir);
+}
+
+int
+main(int argc, const char *argv[])
+{
+    std::string tracedir = gather_trace("", "futex");
+
+    // Now walk the trace and ensure it has futex markers.
+    void *dr_context = dr_standalone_init();
+
+    scheduler_t scheduler;
+    std::vector<scheduler_t::input_workload_t> sched_inputs;
+    sched_inputs.emplace_back(tracedir);
+    if (scheduler.init(sched_inputs, 1, scheduler_t::make_scheduler_serial_options()) !=
+        scheduler_t::STATUS_SUCCESS) {
+        std::cerr << "Failed to initialize scheduler " << scheduler.get_error_string()
+                  << "\n";
+    }
+
+    auto *stream = scheduler.get_stream(0);
+    memref_t memref;
+    int arg_ord = 0;
+    bool saw_maybe_blocking = false;
+    for (scheduler_t::stream_status_t status = stream->next_record(memref);
+         status != scheduler_t::STATUS_EOF; status = stream->next_record(memref)) {
+        assert(status == scheduler_t::STATUS_OK);
+        if (memref.marker.type != TRACE_TYPE_MARKER)
+            continue;
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL) {
+            saw_maybe_blocking = true;
+        }
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_FUNC_ID) {
+            assert(memref.marker.marker_value ==
+                   TRACE_FUNC_ID_SYSCALL_BASE +
+                       IF_X64_ELSE(SYS_futex, SYS_futex & 0xffff));
+        }
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_FUNC_ARG) {
+            switch (arg_ord) {
+            case 0:
+                assert(memref.marker.marker_value ==
+                       reinterpret_cast<uintptr_t>(&futex_var));
+                break;
+            case 1: assert(memref.marker.marker_value == FUTEX_WAKE); break;
+            case 2: assert(memref.marker.marker_value == 1); break;
+            default: assert(memref.marker.marker_value == 0); break;
+            }
+            ++arg_ord;
+        }
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_FUNC_RETVAL) {
+            // Should have succeeded.
+            assert(memref.marker.marker_value == 1);
+        }
+    }
+    assert(saw_maybe_blocking);
+    static constexpr int FUTEX_ARG_COUNT = 6;
+    assert(arg_ord == FUTEX_ARG_COUNT);
```

Comment:
(This assumes only one futex call in the trace. Okay I guess.)

Reply:
If glibc startup adds a futex call we'll see the failure so it won't be silent.

Reply:
Added a comment.